test/integration/consul-container/test/envoy_extensions/wasm_test.go

```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that header is present after wasm applied
+	c2 := cleanhttp.DefaultClient()
+
+	fail := true
+	for i := range [3]int{} {
+		res2, err := c2.Get(fmt.Sprintf("http://localhost:%d", port))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if key := res2.Header.Get("x-test"); key != "" {
+			fmt.Printf("attempt %d: test header missing after WASM applied/n", i)
+		} else {
+			fail = false
+			break
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	if fail {
+		t.Fatal(":(")
+	}
+
+}
+
+// TestWASMLocal Summary
+// This test ensures that a WASM extension with basic functionality is executed correctly.
+// The extension takes an incoming request and adds the header "x-test:true"
+func TestWASMLocal(t *testing.T) {
+	t.Parallel()
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	hostWASMDir := fmt.Sprintf("%s/testdata/wasm_test_files", cwd)
+
+	buildWASM(t, hostWASMDir)
+
+	cluster, _, _ := topology.NewCluster(t, &topology.ClusterConfig{
+		NumServers:                1,
+		NumClients:                1,
+		ApplyDefaultProxySettings: true,
+		BuildOpts: &libcluster.BuildOptions{
+			Datacenter:             "dc1",
+			InjectAutoEncryption:   true,
+			InjectGossipEncryption: true,
+		},
+	})
+
+	clientService := createServices(t, cluster)
+	_, port := clientService.GetAddr()
+	_, adminPort := clientService.GetAdminAddr()
+
+	libassert.AssertUpstreamEndpointStatus(t, adminPort, "static-server.default", "HEALTHY", 1)
+	libassert.GetEnvoyListenerTCPFilters(t, adminPort)
+
+	libassert.AssertContainerState(t, clientService, "running")
+	libassert.AssertFortioName(t, fmt.Sprintf("http://localhost:%d", port), "static-server", "")
+
+	// Check header not present
+	c1 := cleanhttp.DefaultClient()
+	res, err := c1.Get(fmt.Sprintf("http://localhost:%d", port))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// check that header DOES NOT exist before wasm applied
+	if _, ok := res.Header["x-test"]; ok {
+		t.Fatal("unexpected test header present before WASM applied")
+	}
+
+	// wire up the wasm filter
+	consul := cluster.APIClient(0)
+	defaults := api.ServiceConfigEntry{
+		Kind:     api.ServiceDefaults,
+		Name:     "static-server",
+		Protocol: "http",
+		EnvoyExtensions: []api.EnvoyExtension{{
+			Name: "builtin/wasm",
+			Arguments: map[string]any{
+				"Protocol":     "http",
+				"ListenerType": "inbound",
+				"PluginConfig": map[string]any{
+					"VmConfig": map[string]any{
+						"Code": map[string]any{
+							"Local": map[string]any{
+								"Filename": "/wasm_add_header.wasm",
+							},
+						},
+						"Configuration": "plugin configuration",
+					},
+				},
+			},
+		}},
+	}
+
+	_, _, err = consul.ConfigEntries().Set(&defaults, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that header is present after wasm applied
+	c2 := cleanhttp.DefaultClient()
+
+	// The wasm plugin is not always applied on the first call. Retry and see if it is loaded.
+	fail := true
+	for i := range [5]int{} {
```

Comment:
tip: we can use the `retry` library for doing retries, for example: https://github.com/hashicorp/consul/blob/main/api/catalog_test.go#L24-L32